pkg/meta/tkv_badger.go

```diff
@@ -0,0 +1,276 @@
+		value, _ := item.ValueCopy(nil)
+		handler(it.Item().Key(), value)
+	}
+}
+
+func (tx *badgerTxn) get(key []byte) []byte {
+	item, err := tx.t.Get(key)
+	if err == badger.ErrKeyNotFound {
+		return nil
+	}
+	var value []byte
+	err = item.Value(func(val []byte) error {
+		value = append([]byte{}, val...)
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	return value
+}
+
+func (tx *badgerTxn) gets(keys ...[]byte) [][]byte {
+	values := make([][]byte, len(keys))
+	ret := make(map[string][]byte)
+	for _, key := range keys {
+		item, err := tx.t.Get(key)
+		if err == badger.ErrKeyNotFound {
+			ret[string(key)] = []byte{}
+		}
+		if err != nil {
+			panic(err)
+		}
+		err = item.Value(func(val []byte) error {
+			ret[string(key)] = append([]byte{}, val...)
+			return nil
+		})
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	for i, key := range keys {
+		values[i] = ret[string(key)]
+	}
+	return values
+}
+
+func (tx *badgerTxn) Iter(startKey []byte, upperBound []byte) (*badger.Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	it := tx.t.NewIterator(opts)
+	it.Seek(startKey)
+	return it, nil
+}
+
+func (tx *badgerTxn) scanRange0(begin, end []byte, filter func(k, v []byte) bool) map[string][]byte {
+	opts := badger.DefaultIteratorOptions
+	it := tx.t.NewIterator(opts)
+	defer it.Close()
+	var ret = make(map[string][]byte)
+	for it.Seek(begin); it.Valid(); it.Next() {
+		item := it.Item()
+		if bytes.Compare(item.Key(), end) >= 0 {
+			break
+		}
+		key := item.Key()
+		var value []byte
+		err := item.Value(func(val []byte) error {
+			value = append([]byte{}, val...)
+			return nil
+		})
+		if err != nil {
+			panic(err)
+		}
+		if filter == nil || filter(key, value) {
+			ret[string(key)] = value
+		}
+	}
+	return ret
+}
+
+func (tx *badgerTxn) scanRange(begin, end []byte) map[string][]byte {
+	return tx.scanRange0(begin, end, nil)
+}
+
+func (tx *badgerTxn) nextKey(key []byte) []byte {
+	if len(key) == 0 {
+		return nil
+	}
+	next := make([]byte, len(key))
+	copy(next, key)
+	p := len(next) - 1
+	for {
+		next[p]++
+		if next[p] != 0 {
+			break
+		}
+		p--
+		if p < 0 {
+			panic("can't scan keys for 0xFF")
+		}
+	}
+	return next
+}
+
+func (tx *badgerTxn) scanKeys(prefix []byte) [][]byte {
+	endKey := tx.nextKey(prefix)
+	opts := badger.DefaultIteratorOptions
+	it := tx.t.NewIterator(opts)
+
+	defer it.Close()
+	var ret [][]byte
+	for it.Seek(prefix); it.Valid(); it.Next() {
+		item := it.Item()
+		if bytes.Compare(item.Key(), endKey) >= 0 {
+			break
+		}
+		ret = append(ret, item.Key())
+	}
+	return ret
+}
+
+func (tx *badgerTxn) scanValues(prefix []byte, filter func(k, v []byte) bool) map[string][]byte {
+	return tx.scanRange0(prefix, tx.nextKey(prefix), filter)
+}
+
+func (tx *badgerTxn) exist(prefix []byte) bool {
+	nextKey := tx.nextKey(prefix)
+	it, err := tx.Iter(prefix, nextKey)
+	if err != nil {
+		panic(err)
+	}
+	defer it.Close()
+	if it.Valid() {
+		item := it.Item()
+		if bytes.Compare(item.Key(), nextKey) >= 0 {
```

Comment:
Iter() should return any key less than `nextKey`, so we don't need this check, right?